miner/src/main/java/fr/raksrinana/channelpointsminer/miner/database/DatabaseEventHandler.java

```diff
@@ -0,0 +1,155 @@
+				for(var predictor : outcome.getTopPredictors()){
+					database.addUserPrediction(predictor.getUserDisplayName(), streamerUsername, badge);
+				}
+			}
+		}
+		else if(predictionEvent.getStatus() == EventStatus.CANCELED){
+			log.info("Prediction-Update: Event CANCELED. Streamer: {}, Title: {}", streamerUsername, predictionEvent.getTitle());
+			
+			var ended = Optional.ofNullable(predictionEvent.getEndedAt()).map(ZonedDateTime::toInstant).orElseGet(TimeFactory::now);
+			database.cancelPrediction(predictionEvent.getId(), predictionEvent.getChannelId(), predictionEvent.getTitle(), predictionEvent.getCreatedAt().toInstant(), ended);
+		}
+		else if(predictionEvent.getStatus() == EventStatus.RESOLVED){
+			var winningOutcomeId = predictionEvent.getWinningOutcomeId();
+			var winningOutcome = predictionEvent.getOutcomes().stream()
+					.filter(e -> e.getId().equals(winningOutcomeId))
+					.findAny().orElseThrow();
+			var winningOutcomeBadge = winningOutcome.getBadge().getVersion();
+			
+			log.info("Prediction-Update: Event RESOLVED. Streamer: {}, Title: {}, Outcome: {}", streamerUsername, predictionEvent.getTitle(), winningOutcome.getTitle());
+			
+			var ended = Optional.ofNullable(predictionEvent.getEndedAt()).map(ZonedDateTime::toInstant).orElseGet(TimeFactory::now);
+			var totalPoints = predictionEvent.getOutcomes().stream().mapToDouble(Outcome::getTotalPoints).sum();
+			var returnRatio = totalPoints / winningOutcome.getTotalPoints();
+			database.resolvePrediction(predictionEvent.getId(), predictionEvent.getChannelId(), predictionEvent.getTitle(), predictionEvent.getCreatedAt().toInstant(), ended, winningOutcome.getTitle(), winningOutcomeBadge, returnRatio);
+		}
+	}
+	
+	@Override
+	public void onPointsEarnedEvent(@NotNull PointsEarnedEvent event) throws SQLException{
+		var pointsEarnedData = event.getPointsEarnedData();
+		var reasonCode = pointsEarnedData.getPointGain().getReasonCode();
+		updateBalance(event, pointsEarnedData.getBalance(), reasonCode.name());
+	}
+	
+	@Override
+	public void onPointsSpentEvent(@NotNull PointsSpentEvent event) throws SQLException{
+		updateBalance(event, event.getPointsSpentData().getBalance(), null);
+	}
+	
+	@Override
+	public void onPredictionMadeEvent(@NotNull PredictionMadeEvent event) throws SQLException{
+		var placedPrediction = event.getPlacedPrediction();
+		addPrediction(event, placedPrediction.getEventId(), "PREDICTED", Integer.toString(placedPrediction.getAmount()));
+		var outcomeId = event.getPlacedPrediction().getOutcomeId();
+		if(event.getPlacedPrediction().getBettingPrediction() != null && event.getStreamerUsername().isPresent()){
+			var placedOutcome = event.getPlacedPrediction().getBettingPrediction()
+					.getEvent().getOutcomes().stream().filter(o -> o.getId().equals(outcomeId)).findFirst();
+			if(placedOutcome.isPresent()){
+				database.addUserPrediction(event.getMiner().getUsername(), event.getStreamerUsername().get(), placedOutcome.get().getBadge().getVersion());
```

Comment:
I wonder, is it useful to record our own predictions? Like we'll never use them to know if we're a trustworthy user as we'll always place the bet after computing trust scores.

Reply:
Yeah, they aren't useful for the outcome picker. I added it to be able to better compare oneself to other people (of course the information currently isn't visible without connecting to the db). I would leave it in, as it is basically free.